corehq/apps/api/user_updates.py

```diff
@@ -0,0 +1,135 @@
+
+    if user_change_logger:
+        user_change_logger.add_change_message(UserChangeMessage.password_reset())
+
+
+def _update_default_phone_number(user, phone_number, user_change_logger):
+    old_phone_numbers = set(user.phone_numbers)
+    if not isinstance(phone_number, str):
+        raise InvalidFormatException('string')
+    formatted_phone_number = strip_plus(phone_number)
+    user.set_default_phone_number(formatted_phone_number)
+
+    if user_change_logger:
+        numbers_added, numbers_removed = find_differences_in_list(
+            target=[phone_number],
+            source=list(old_phone_numbers)
+        )
+
+        change_messages = {}
+        if numbers_added:
+            change_messages.update(
+                UserChangeMessage.phone_numbers_added(list(numbers_added))["phone_numbers"]
+            )
+            user_change_logger.add_change_message({'phone_numbers': change_messages})
+
+
+def _update_phone_numbers(user, phone_numbers, user_change_logger):
+    old_phone_numbers = set(user.phone_numbers)
+    new_phone_numbers = set()
+    user.phone_numbers = []
+    for idx, phone_number in enumerate(phone_numbers):
+        formatted_phone_number = strip_plus(phone_number)
+        new_phone_numbers.add(formatted_phone_number)
+        user.add_phone_number(formatted_phone_number)
+        if idx == 0:
+            user.set_default_phone_number(formatted_phone_number)
+
+    if user_change_logger:
+        (numbers_added, numbers_removed) = find_differences_in_list(
+            target=list(new_phone_numbers),
+            source=list(old_phone_numbers)
+        )
+
+        change_messages = {}
+        if numbers_removed:
+            change_messages.update(
+                UserChangeMessage.phone_numbers_removed(list(numbers_removed))["phone_numbers"]
+            )
+
+        if numbers_added:
+            change_messages.update(
+                UserChangeMessage.phone_numbers_added(list(numbers_added))["phone_numbers"]
+            )
+
+        if change_messages:
+            user_change_logger.add_change_message({'phone_numbers': change_messages})
+
+
+def _update_groups(user, group_ids, user_change_logger):
+    groups_updated = user.set_groups(group_ids)
+    if user_change_logger and groups_updated:
+        groups = []
+        if group_ids:
+            groups = [Group.wrap(doc) for doc in get_docs(Group.get_db(), group_ids)]
+        user_change_logger.add_info(UserChangeMessage.groups_info(groups))
+
+
+def _update_user_data(user, user_data, user_change_logger):
+    original_user_data = user.metadata.copy()
+    user.update_metadata(user_data)
+
+    if user_change_logger and original_user_data != user.user_data:
+        user_change_logger.add_changes({'user_data': user.user_data})
+
+
+def _simple_update(user, key, value, user_change_logger):
+    if user_change_logger and getattr(user, key) != value:
+        user_change_logger.add_changes({key: value})
+    setattr(user, key, value)
```

Comment:
should this be under the `if` statement? Functionally it's the same but since we're already checking for a change it makes logical sense to only set the attribute if it's different.

Reply:
Yeah my reason for keeping it separate was preferring that over this
```
if getattr(user, key) != value:
  setattr(user, key, value)
  if user_change_logger:
    user_change_logger.add_changes({key: value})
```
but I agree if we are going to check for a diff, might as well only set it if there is a diff.

Reply:
oh right, I glossed over that the user_change_logger was also part of the if condition

Reply:
I changed it anyway. See 4c3fca64ef3a63192091a12dcb12490f35e98180